// @nx/angular/src/generators/component/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const devkit_1 = require("@nx/devkit");
const path_1 = require("../../utils/path");
const selector_1 = require("../../utils/selector");
function normalizeOptions(tree, options) {
    options.type ??= 'component';
    const { directory, filePath, name, path, root, sourceRoot } = (0, path_1.normalizeNameAndPaths)(tree, options);
    const { prefix } = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    const selector = options.selector ??
        (0, selector_1.buildSelector)(tree, name, options.prefix, prefix, 'fileName');
    return {
        ...options,
        name,
        changeDetection: options.changeDetection ?? 'Default',
        style: options.style ?? 'css',
        flat: options.flat ?? false,
        directory,
        filePath,
        path,
        projectSourceRoot: sourceRoot,
        projectRoot: root,
        selector,
    };
}
exports.normalizeOptions = normalizeOptions;
