// @nx/workspace/src/utils/workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateBuilderConfig = exports.updateWorkspace = exports.getWorkspace = void 0;
const core_1 = require("@angular-devkit/core");
function createHost(tree) {
    return {
        async readFile(path) {
            const data = tree.read(path);
            if (!data) {
                throw new Error('File not found.');
            }
            return data.toString();
        },
        async writeFile(path, data) {
            return tree.overwrite(path, data);
        },
        async isDirectory(path) {
            // approximate a directory check
            // special case needed when testing wrapped schematics
            if (path === '/')
                return true;
            return !tree.exists(path) && tree.getDir(path).subfiles.length > 0;
        },
        async isFile(path) {
            return tree.exists(path);
        },
    };
}
/**
 * @deprecated This will be removed in v17. Prefer writing Nx Generators with @nx/devkit. This function can be replaced with 'getProjects' with Nx Devkit.
 */
async function getWorkspace(tree, path = '/') {
    const host = createHost(tree);
    const { workspace } = await core_1.workspaces.readWorkspace(path, host);
    return workspace;
}
exports.getWorkspace = getWorkspace;
/**
 * @deprecated This will be removed in v17. Prefer writing Nx Generators with @nx/devkit. This function can be replaced with 'updateProjectConfiguration' with Nx Devkit.
 */
function updateWorkspace(updaterOrWorkspace) {
    return async (tree) => {
        const host = createHost(tree);
        if (typeof updaterOrWorkspace === 'function') {
            const { workspace } = await core_1.workspaces.readWorkspace('/', host);
            const result = updaterOrWorkspace(workspace);
            if (result !== undefined) {
                await result;
            }
            await core_1.workspaces.writeWorkspace(workspace, host);
        }
        else {
            await core_1.workspaces.writeWorkspace(updaterOrWorkspace, host);
        }
    };
}
exports.updateWorkspace = updateWorkspace;
/**
 * Updates builder options for options and configurations for given builder names
 * @deprecated This will be removed in v17. Prefer writing Nx Generators with @nx/devkit. This function can be replaced with 'forEachExecutorOptions' with Nx Devkit.
 */
function updateBuilderConfig(updater, ...builderNames) {
    return updateWorkspace((workspace) => {
        if (!workspace.projects) {
            return;
        }
        workspace.projects.forEach((project) => {
            project.targets.forEach((target) => {
                if (!builderNames.includes(target.builder)) {
                    return;
                }
                if (target.options) {
                    target.options = updater(target.options, target, project);
                }
                if (!target.configurations) {
                    return;
                }
                Object.entries(target.configurations).forEach(([configName, options]) => {
                    target.configurations[configName] = updater(options, target, project);
                });
            });
        });
    });
}
exports.updateBuilderConfig = updateBuilderConfig;
