// nx-cloud/lib/utilities/nx-imports-light.js
"use strict";
try {
    try {
        const { output } = require('nx/src/utils/output');
        let workspaceRoot;
        try {
            workspaceRoot = require('nx/src/utils/app-root').workspaceRoot;
        }
        catch (_a) {
            workspaceRoot = require('nx/src/utils/workspace-root').workspaceRoot;
        }
        exports.workspaceRoot = workspaceRoot;
        exports.output = output;
    }
    catch (e) {
        const { output } = require('@nrwl/workspace/src/utilities/output');
        const { appRootPath } = require('@nrwl/tao/src/utils/app-root');
        exports.workspaceRoot = appRootPath;
        exports.output = output;
    }
}
catch (e) {
    const formatMessage = (input) => {
        var _b;
        return `${input.title}\n\n${(_b = input.bodyLines) === null || _b === void 0 ? void 0 : _b.join('\n')}`;
    };
    exports.output = {
        note: (input) => console.info(formatMessage(input)),
        error: (input) => console.error(formatMessage(input)),
        warn: (input) => console.warn(formatMessage(input)),
        success: (input) => console.log(formatMessage(input)),
        addVerticalSeparator: () => '',
        addNewline: () => console.log('\n'),
    };
    exports.workspaceRoot = process.cwd();
}
